src/plugin.js

```diff
@@ -1,115 +1,78 @@
-    snapshotResults.diffOutputPath = diffDotPath;
-
-    return {
-      path: diffDotPath,
-    };
-  }
-
-  fs.copySync(snapshotKebabPath, snapshotDotPath);
-  fs.removeSync(snapshotKebabPath);
-  snapshotResults.diffOutputPath = snapshotDotPath;
-
-  return {
-    path: snapshotDotPath,
+import { COMPARE } from './constants';
+import compare from './compare';
+
+let screenshotDetails = {};
+
+const matchImageSnapshotPlugin = (name, options, config) => {
+    // Check if we should skip the comparison, for example while debugging a case.
+    if (options.skipCompare.call()) {
+      return Promise.resolve('Snapshot Compare skipped.');
+    }
+
+    // Set all paths needed in our plugin.
+    const relativePath = process.cwd();
+    const screenshotPath = screenshotDetails.path.replace(
+      config.screenshotsFolder,
+      ''
+    );
+    const screenshotFolder = screenshotDetails.path
+      .replace(relativePath, '')
+      .replace(screenshotPath, '');
+
+    const diffPath = path.join(
+      relativePath,
+      options.diffFolder || screenshotFolder,
+      screenshotPath.replace('.png', options.dotDiff)
+    );
+
+    const snapPath = path.join(
+      relativePath,
+      options.snapshotFolder || screenshotFolder,
+      screenshotPath.replace('.png', options.dotSnap)
+    );
+
+    return compare(snapPath, screenshotDetails.path, diffPath, config.env.updateSnapshots || false).then(({ diff, total, msg }) => {
+        options.onDiffFinished.call(snapPath, screenshotDetails.path, diffPath);
+
+        if (msg) {
+          return msg;
+        }
+
+        if (options.thresholdType === 'pixel') {
+            if(diff > options.threshold) {
+                throw new Error(`Image comparison failed, the change of ${diff} is higher than the allowed ${options.threshold} pixels.`);
+            }
+        } else if (options.threshold === 'percentage') {
+            const percentage = ((total - diff) / total) * 100;
+
+            if(percentage > options.threshold) {
+                throw new Error(`Image comparison failed, the change of ${diff} is higher than the allowed ${options.threshold} percentage.`);
+            }
+        }
+
+        return 'Screenshot matched';
+    });
   };
-}
 
-export function addMatchImageSnapshotPlugin(on) {
+const addMatchImageSnapshotPlugin = (on, config, pluginOptions) => {
+  const pluginDefaults = {
+    dotSnap: '.snap.png',
+    dotDiff: '.diff.png',
+    snapshotFolder: undefined,
+    diffFolder: undefined,
+    skipCompare: () => false,
+    threshold: 0,
+    thresholdType: 'pixel',
+    onDiffFinished: () => {}
+  };
   on('task', {
-    [MATCH]: matchImageSnapshotOptions,
-    [RECORD]: matchImageSnapshotResults,
+    [COMPARE]: (name, options) =>
+      matchImageSnapshotPlugin(name, { ...pluginDefaults, ...pluginOptions, ...options }, config),
   });
-  on('after:screenshot', matchImageSnapshotPlugin);
-}
+  on('after:screenshot', details => {
+    screenshotDetails = details;
```

Comment:
The reason we have two tasks is so that we can [return the updated image path](https://github.com/palmerhq/cypress-image-snapshot/blob/master/src/plugin.js#L105) from the `after:screenshot` callback.  The Cypress docs [have a section explaining this](https://docs.cypress.io/api/plugins/after-screenshot-api.html), but the main we return the updated path is so that Cypress can upload the images and/or composite diffs to the dashboard.

I am not sure it is possible to maintain this functionality with only one command.

Reply:
Was not aware of that feature, let me have a look at that and update the code if needed.

Reply:
@crosscompile if I understand correctly we return the new path in two scenarios. If we update the snapshot file and when there is a diff, correct?

Reply:
We return the [path to the diff](https://github.com/palmerhq/cypress-image-snapshot/blob/master/src/plugin.js#L96) if the comparison fails and the [path to the snapshot](https://github.com/palmerhq/cypress-image-snapshot/blob/master/src/plugin.js#L105) if it passes.

Since the original screenshot path is always modified, we need to return the new path if we want any images to be uploaded to the dashboard.

Reply:
I think one way I can see to do this would be to append a special key to the filename when calling `cy.screenshot` (so that we know the screenshot path needs to be updated) and then removing the ability to [pass custom options](https://github.com/palmerhq/cypress-image-snapshot/blob/master/src/plugin.js#L36-L46) aside from the screenshot options to each individual `matchImageSnapshot` command.

This tradeoff might be worth it.

Reply:
But if we are not able to provide custom options to the matchImageSnapshot, wouldn't this mean we can't set the options like `failureThreshold`? These kind of options are relatively important to us actually.

We are using screenshot comparison to test our component library, in for example a `input` field the blinking caret can cause flaky cases, so this threshold helps us work around that.